pkg/cmd/staticpodcontroller/staticpodcontroller.go

```diff
@@ -0,0 +1,301 @@
+	if err != nil {
+		return err
+	}
+	clientmc, err := mcfgclientset.NewForConfig(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	kubeInformerFactory := informers.NewFilteredSharedInformerFactory(clientset, 0, "openshift-etcd", nil)
+	nodeName := os.Getenv("NODE_NAME")
+	var localEtcdName string
+	if nodeName != "" {
+		localEtcdName = fmt.Sprintf("etcd-member-%s", nodeName)
+	}
+
+	staticPodController := NewStaticPodController(
+		kubeInformerFactory,
+		clientset,
+		clientmc,
+		localEtcdName,
+		// ctx.EventRecorder,
+	)
+
+	kubeInformerFactory.Start(ctx.Done())
+
+	go staticPodController.Run(ctx.Done())
+
+	<-ctx.Done()
+	cancel()
+	return fmt.Errorf("stopped")
+}
+
+type StaticPodController struct {
+	podInformer                            corev1informer.SecretInformer
+	kubeInformersForOpenshiftEtcdNamespace cache.SharedIndexInformer
+	clientset                              corev1client.Interface
+	clientmc                               mcfgclientset.Interface
+	localEtcdName                          string
+
+	cachesToSync  []cache.InformerSynced
+	queue         workqueue.RateLimitingInterface
+	eventRecorder events.Recorder
+}
+
+func NewStaticPodController(
+	kubeInformersForOpenshiftEtcdNamespace informers.SharedInformerFactory,
+	clientset corev1client.Interface,
+	clientmc mcfgclientset.Interface,
+	localEtcdName string,
+	// eventRecorder events.Recorder,
+) *StaticPodController {
+	c := &StaticPodController{
+		// eventRecorder: eventRecorder.WithComponentSuffix("static-pod-controller"),
+		clientset:     clientset,
+		clientmc:      clientmc,
+		localEtcdName: localEtcdName,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "StaticPodController"),
+	}
+	kubeInformersForOpenshiftEtcdNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+	// kubeInformersForOpenshiftEtcdNamespace.Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *StaticPodController) sync() error {
+	pod, err := c.clientset.CoreV1().Pods("openshift-etcd").Get(c.localEtcdName, metav1.GetOptions{})
+	if err != nil {
+		klog.Infof("No Pod found in openshift-etcd with name %s", c.localEtcdName)
+		return err
+	}
+	staticPodPath := fmt.Sprintf("%s/%s", manifestDir, etcdMemberFileName)
+	// If last status is CrashLoopBackOff we perform further inspection to verify.
+	if pod.Status.ContainerStatuses[0].State.Waiting != nil && pod.Status.ContainerStatuses[0].State.Waiting.Reason == "CrashLoopBackOff" {
+
+		klog.Warningf("%s is unhealthy", c.localEtcdName)
+
+		// if the data-dir is missing we are going to give it some time to recover.
+		if _, err := os.Stat(fmt.Sprintf("%s/member/snap", dataDir)); os.IsNotExist(err) {
+			klog.Warningf("data-dir already removed %s", dataDir)
+			return nil
+		}
+
+		if c.isContainerCrashLoop("etcd-member") {
+			etcdMember, err := c.getMachineConfigData(staticPodPath, "master")
```

Comment:
we need to move this over to observation so that we get a more accurate picture of the crashloop. here we need to worry about if we can see `MemberRemove` or `MemberDegrade`